plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/RetryingJdbcClient.java

```diff
@@ -0,0 +1,527 @@
+import java.util.Set;
+
+import static io.trino.plugin.jdbc.RetryingModule.retry;
+import static java.util.Objects.requireNonNull;
+
+public class RetryingJdbcClient
+        implements JdbcClient
+{
+    private final JdbcClient delegate;
+    private final RetryPolicy<Object> policy;
+
+    @Inject
+    public RetryingJdbcClient(@ForRetrying JdbcClient delegate, RetryPolicy<Object> policy)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.policy = requireNonNull(policy, "policy is null");
+    }
+
+    @Override
+    public boolean schemaExists(ConnectorSession session, String schema)
+    {
+        return retry(policy, () -> delegate.schemaExists(session, schema));
+    }
+
+    @Override
+    public Set<String> getSchemaNames(ConnectorSession session)
+    {
+        return retry(policy, () -> delegate.getSchemaNames(session));
+    }
+
+    @Override
+    public List<SchemaTableName> getTableNames(ConnectorSession session, Optional<String> schema)
+    {
+        return retry(policy, () -> delegate.getTableNames(session, schema));
+    }
+
+    @Override
+    public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        return retry(policy, () -> delegate.getTableHandle(session, schemaTableName));
+    }
+
+    @Override
+    public JdbcTableHandle getTableHandle(ConnectorSession session, PreparedQuery preparedQuery)
+    {
+        return retry(policy, () -> delegate.getTableHandle(session, preparedQuery));
+    }
+
+    @Override
+    public JdbcProcedureHandle getProcedureHandle(ConnectorSession session, ProcedureQuery procedureQuery)
+    {
+        return retry(policy, () -> delegate.getProcedureHandle(session, procedureQuery));
+    }
+
+    @Override
+    public List<JdbcColumnHandle> getColumns(ConnectorSession session, JdbcTableHandle tableHandle)
+    {
+        return retry(policy, () -> delegate.getColumns(session, tableHandle));
+    }
+
+    @Override
+    public Iterator<RelationColumnsMetadata> getAllTableColumns(ConnectorSession session, Optional<String> schema)
+    {
+        return retry(policy, () -> delegate.getAllTableColumns(session, schema));
+    }
+
+    @Override
+    public List<RelationCommentMetadata> getAllTableComments(ConnectorSession session, Optional<String> schema)
+    {
+        return retry(policy, () -> delegate.getAllTableComments(session, schema));
+    }
+
+    @Override
+    public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    {
+        // no retrying as it could be not idempotent operation (connection could be not reusable after the first failure)
+        return delegate.toColumnMapping(session, connection, typeHandle);
+    }
+
+    @Override
+    public List<ColumnMapping> toColumnMappings(ConnectorSession session, List<JdbcTypeHandle> typeHandles)
+    {
+        // there should be no remote database interaction
```

Comment:
not exactly. I think the reason we don't retry here is because it ends up calling `toColumnMapping` which we cannot retry anyway.

Reply:
Added retrying here too. Thanks for the feedback.